File: clients/CaptureCore/Sources/CaptureCore/Schema.swift
import Foundation
import PowerSync

public let TASKS_TABLE = "tasks"
public let TAGS_TABLE = "tags"
public let CATEGORIES_TABLE = "categories"
public let TASK_EVENTS_TABLE = "task_events"
public let TASK_ATTACHMENTS_TABLE = "task_attachments"
public let AGENT_PROPOSALS_TABLE = "agent_proposals"

/// Client-side SQLite schema. Mirrors Postgres `public.tasks` / `public.tags`. The `id`
/// column is auto-created by PowerSync — never declare it.
public let AppSchema = Schema(
    Table(
        name: TASKS_TABLE,
        columns: [
            .text("owner_id"),
            .text("parent_task_id"),
            .text("title"),
            .text("notes"),
            .text("status"),
            .text("category"),
            .text("tags"),               // JSON array of tag names
            .text("due_at"),
            .integer("priority"),
            .text("github_repo"),
            .text("github_url"),
            .text("suggested_due_at"),
            .text("suggested_category"),
            .real("suggestion_confidence"),
            .text("suggestion_source"),
            .text("source"),
            .text("created_at"),
            .text("updated_at"),
            .text("confirmed_at"),
            .text("completed_at")
        ],
        indexes: [
            Index(name: "by_status", columns: [IndexedColumn.ascending("status")]),
            Index(name: "by_parent", columns: [
                IndexedColumn.ascending("parent_task_id"),
                IndexedColumn.ascending("status")
            ])
        ]
    ),
    Table(
        name: TAGS_TABLE,
        columns: [
            .text("owner_id"),
            .text("name"),
            .text("color"),
            .text("created_at"),
            .text("updated_at")
        ]
    ),
    Table(
        name: CATEGORIES_TABLE,
        columns: [
            .text("owner_id"),
            .text("name"),
            .text("color"),
            .text("created_at"),
            .text("updated_at")
        ]
    ),
    Table(
        name: TASK_EVENTS_TABLE,
        columns: [
            .text("owner_id"),
            .text("task_id"),
            .text("actor"),
            .text("event_type"),
            .text("title"),
            .text("body"),
            .text("metadata"),
            .text("created_at")
        ],
        indexes: [
            Index(name: "by_task_created", columns: [
                IndexedColumn.ascending("task_id"),
                IndexedColumn.ascending("created_at")
            ])
        ]
    ),
    Table(
        name: TASK_ATTACHMENTS_TABLE,
        columns: [
            .text("owner_id"),
            .text("task_id"),
            .text("filename"),
            .text("mime_type"),
            .integer("byte_size"),
            .text("preview_data_url"),
            .text("created_at")
        ],
        indexes: [
            Index(name: "by_task_created", columns: [
                IndexedColumn.ascending("task_id"),
                IndexedColumn.ascending("created_at")
            ])
        ]
    ),
    Table(
        name: AGENT_PROPOSALS_TABLE,
        columns: [
            .text("owner_id"),
            .text("task_id"),
            .text("proposal_type"),
            .text("status"),
            .text("title"),
            .text("body"),
            .text("payload"),
            .text("provenance"),
            .real("confidence"),
            .text("source"),
            .text("created_at"),
            .text("updated_at"),
            .text("decided_at"),
            .text("applied_at")
        ],
        indexes: [
            Index(name: "by_status", columns: [
                IndexedColumn.ascending("status"),
                IndexedColumn.ascending("created_at")
            ]),
            Index(name: "by_task_status", columns: [
                IndexedColumn.ascending("task_id"),
                IndexedColumn.ascending("status")
            ])
        ]
    )
)

File: web/src/powersync/schema.ts
import { column, Schema, Table } from '@powersync/web';

// Client-side SQLite schema (mirrors Postgres public.tasks). `id` is implicit.
// Timestamps are stored as ISO-8601 text; PowerSync column types are text/integer/real only.
const tasks = new Table(
  {
    owner_id: column.text,
    parent_task_id: column.text,
    title: column.text,
    notes: column.text,
    status: column.text,
    category: column.text,
    tags: column.text,
    due_at: column.text,
    priority: column.integer,
    github_repo: column.text,
    github_url: column.text,
    suggested_due_at: column.text,
    suggested_category: column.text,
    suggestion_confidence: column.real,
    suggestion_source: column.text,
    source: column.text,
    created_at: column.text,
    updated_at: column.text,
    confirmed_at: column.text,
    completed_at: column.text
  },
  { indexes: { by_status: ['status', 'created_at'], by_parent: ['parent_task_id', 'status', 'created_at'] } }
);

const tags = new Table(
  {
    owner_id: column.text,
    name: column.text,
    color: column.text,
    created_at: column.text,
    updated_at: column.text
  },
  { indexes: {} }
);

const categories = new Table(
  {
    owner_id: column.text,
    name: column.text,
    color: column.text,
    created_at: column.text,
    updated_at: column.text
  },
  { indexes: {} }
);

const categorisation_rules = new Table(
  {
    owner_id: column.text,
    title: column.text,
    instructions: column.text,
    category: column.text,
    tags: column.text,
    enabled: column.integer,
    created_at: column.text,
    updated_at: column.text
  },
  { indexes: {} }
);

const task_events = new Table(
  {
    owner_id: column.text,
    task_id: column.text,
    actor: column.text,
    event_type: column.text,
    title: column.text,
    body: column.text,
    metadata: column.text,
    created_at: column.text
  },
  { indexes: { by_task_created: ['task_id', 'created_at'] } }
);

const task_attachments = new Table(
  {
    owner_id: column.text,
    task_id: column.text,
    filename: column.text,
    mime_type: column.text,
    byte_size: column.integer,
    preview_data_url: column.text,
    created_at: column.text
  },
  { indexes: { by_task_created: ['task_id', 'created_at'] } }
);

const agent_proposals = new Table(
  {
    owner_id: column.text,
    task_id: column.text,
    proposal_type: column.text,
    status: column.text,
    title: column.text,
    body: column.text,
    payload: column.text,
    provenance: column.text,
    confidence: column.real,
    source: column.text,
    created_at: column.text,
    updated_at: column.text,
    decided_at: column.text,
    applied_at: column.text
  },
  { indexes: { by_status: ['status', 'created_at'], by_task_status: ['task_id', 'status'] } }
);

export const AppSchema = new Schema({ tasks, tags, categories, categorisation_rules, task_events, task_attachments, agent_proposals });

export type Database = (typeof AppSchema)['types'];
export type TaskRecord = Database['tasks'];
export type TagRecord = Database['tags'];
export type CategoryRecord = Database['categories'];
export type CategorisationRuleRecord = Database['categorisation_rules'];
export type TaskEventRecord = Database['task_events'];
export type TaskAttachmentRecord = Database['task_attachments'];
export type AgentProposalRecord = Database['agent_proposals'];
